jscomp/core/lam_analysis.ml

```diff
@@ -32,7 +32,7 @@ let not_zero_constant (x : Lam_constant.t) =
 let rec no_side_effects (lam : Lam.t) : bool =
   match lam with
   | Lvar _ | Lconst _ | Lfunction _ -> true
-  | Lglobal_module _ -> true
+  | Lglobal_module (_, dynamic_import) -> not dynamic_import
```

Comment:
What is the impact of this change in practice?
Do we have tests that show what would happen with and without this change?

Reply:
Good catch. Now that I look at it again, Lglobal_module always being no side effect might have a different context. I'll do some testing.

Reply:
I just found there are already `side_effect.res` and `side_effect_free.res` in tests. I've changed tests to `import_side_effect.res` and `import_side_effect_free.res`. I confirmed that the compiler correctly detects side effects when dynamically importing a value with no side effects. So, it seems that the original Lglobal_module expression without side effects is indeed correct in the existing code.

Reply:
See https://github.com/rescript-lang/rescript-compiler/pull/6232/commits/b5d836616bf5ea652d90904945eb35a59e6e5de6

Reply:
The comment is indeed correct. https://github.com/rescript-lang/rescript-compiler/blob/master/jscomp/core/lam_analysis.ml#L35-L38